Snakefile

```diff
@@ -1104,6 +1116,57 @@ rule learn_sentencepiece:
         "--normalization_rule_name nmt_nfkc_cf"
 
 
+# Take another (larger) sample of the full data set previously sampled to train
+# SentencePiece, and encode it as ids using the trained SentencePiece model.
+# Split the encoded output into files containing uniform numbers of ids.
+rule encode_sentencepiece_ids:
+    input:
+        "data/monolingual/{lang}.txt.gz",
+        "data/sentencepiece/{lang}.model"
+    output:
+        temp(expand("data/sentencepiece/{{lang}}.spm_txt_ids/length_{n}.txt",
+                    n=range(MAX_SPM_ENCODE_IDS + 1)))
+    run:
+        # We invoke awk twice.  The first time it samples the input data file.
+        # The second time it distributes the output of spm_encode to output
+        # files whose names include the length of each encoded input line
+        # (using the awk built in variable NF denoting the number of fields
+        # in a line).
+        data_file, model_file = input
+        output_dir = os.path.dirname(output[0])
+        output_prefix = '"{}"'.format(os.path.join(output_dir, "length_"))
+        output_suffix = '".txt"'
+        length_bound = MAX_SPM_ENCODE_IDS + 1
+        for output_file in output:  # make empty files
+             shell("touch {output_file}")
+        shell(
+            "zcat {data_file} | "
+            "awk 'NR % {RECIPROCAL_OF_FRACTION_OF_DATA_TO_SPM_ENCODE} == 0' | "
+            "spm_encode --model={model_file} --output_format=id | "
+            "awk 'NF < {length_bound} {{print $0 > ({output_prefix} NF {output_suffix})}}'"
```

Comment:
I said before that I'm okay with awk one-liners in the case that you're doing a very well-known thing with awk, such as sampling every Nth line. But this is a more bespoke awk script, and it's especially difficult to interpret given the two different kinds of braces you have to use when embedding it in a Snakefile.

I suggest it should be in its own .awk file, with a comment or two. You can take in `output_prefix` and `length_bound` as parameters. `output_suffix` doesn't have to be a parameter.

Reply:
Yeah, I had made this a separate script even before seeing your comment.